lib/src/lints/flutter/prefer_async_callback.dart

```diff
@@ -37,6 +39,15 @@ class PreferAsyncCallback extends DartLintRule {
       final typeParameters = node.typeParameters?.typeParameters;
       if (typeParameters != null) return;
 
+      final returnTypeArgs =
+          (node.returnType! as NamedType).typeArguments?.arguments;
+      switch (returnTypeArgs) {
+        case [final NamedType type] when type.type is VoidType:
```

Comment:
This is smart! And shall we extract this logic to a separate function and use `if` instead of `switch`?
Perhaps something like this.

```dart
bool _isFutureVoid(TypeAnnotation typeAnnotation) {
  final type = typeAnnotation.type;
  if (type == null) return false;
  if (!type.isDartAsyncFuture) return false;
  if (typeAnnotation is! NamedType) return false;

  final typeArguments = typeAnnotation.typeArguments?.arguments;
  if (typeArguments case [final NamedType type] when type.type is VoidType) {
    return true;
  }
  return false;
}
```

And at the beginning, we do

```dart
final returnType = node.returnType;
if (returnType == null || !_isFutureVoid(returnType)) return;
```

Reply:
> ```dart
>  if (typeAnnotation is! NamedType) return false;
> ```

I don't think we need this case as we already check for the dart's future type case and the Future type will always be a NamedType.

Reply:
And also the if statement is not very readable for this long pattern match expression. we should use switch here for more readability.

```dart
bool _isFutureVoid(TypeAnnotation typeAnnotation) {
  final type = typeAnnotation.type;
  if (type == null || !type.isDartAsyncFuture) return false;

  final typeArgs = (typeAnnotation as NamedType).typeArguments?.arguments;

  return switch (typeArgs) {
    [final NamedType type] when type.type is VoidType => true,
    _ => false,
  };
}
```

Reply:
Final code would look like this


```dart

   context.registry.addGenericFunctionType((node) {
      final returnType = node.returnType;
      if (returnType == null || !_isFutureVoid(returnType)) return;

      final parameters = node.parameters.parameters;
      if (parameters.isNotEmpty) return;

      final typeParameters = node.typeParameters?.typeParameters;
      if (typeParameters != null) return;

      reporter.reportErrorForNode(code, node);
    });

  bool _isFutureVoid(TypeAnnotation typeAnnotation) {
    final type = typeAnnotation.type;
    if (type == null || !type.isDartAsyncFuture) return false;

    final typeArgs = (typeAnnotation as NamedType).typeArguments?.arguments;

    return switch (typeArgs) {
      [final NamedType type] when type.type is VoidType => true,
      _ => false,
    };
  }
```

Reply:
True, using `switch` here is easier to read but I still do not like using `(typeAnnotation as NamedType)`.
It is not easy to spot why it is safe to use `typeAnnotation as NamedType` here.

Shall we do something like this?
```dart
// Since we know it's a Future, we can safely cast it to NamedType.
typeAnnotation as NamedType;
final typeArguments = typeAnnotation.typeArguments?.arguments;
```

Reply:
Sure,